src/main/java/nu/studer/teamcity/buildscan/internal/ArtifactBuildScanDataStore.java

```diff
@@ -0,0 +1,87 @@
+package nu.studer.teamcity.buildscan.internal;
+
+import com.google.common.base.Charsets;
+import jetbrains.buildServer.ArtifactsConstants;
+import jetbrains.buildServer.serverSide.SBuild;
+import nu.studer.teamcity.buildscan.BuildScanReference;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class ArtifactBuildScanDataStore implements BuildScanDataStore {
+
+    private static final Logger LOGGER = Logger.getLogger("jetbrains.buildServer.BUILDSCAN");
+    public static final String BUILDSCANS_DIR = "buildscans";
+    public static final String BUILDSCAN_URLS_FILE = "buildscan_urls.txt";
+    private final ReadOnlyBuildScanDataStore fallbackDataStore;
+
+    public ArtifactBuildScanDataStore(ReadOnlyBuildScanDataStore fallbackDataStore) {
+        this.fallbackDataStore = fallbackDataStore;
+    }
+
+    @Override
+    public void mark(SBuild build) {
+        final Path buildScanFile = getBuildScanFile(build.getArtifactsDirectory());
+        try {
+            createIfNotExists(buildScanFile);
+        } catch (IOException ex) {
+            LOGGER.error(String.format("Could not create buildscan file %s", buildScanFile), ex);
+        }
+    }
+
+    @Override
+    public void store(SBuild build, String buildScanUrl) {
+        final Path buildScanFile = getBuildScanFile(build.getArtifactsDirectory());
+        try {
+            createIfNotExists(buildScanFile);
+            Files.write(buildScanFile, Collections.singletonList(buildScanUrl), StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.APPEND);
```

Comment:
We don't need CREATE here since we already ensured above that the file exists.

Reply:
I removed CREATE on master.

Reply:
Good catch.